### ISSUE TITLE
Incorrect SQL resulted for 3 embedded model merge

### ISSUE THREAD

**Issue by majioa** [rails/rails#25806]
### Steps to reproduce

Create 5 models (4 tables) with has_many relations:

``` ruby
class Organization < ActiveRecord::Base
  has_many :lots, inverse_of: :customer, foreign_key: :customer_id
  scope :by_okpd_code, ->(okpd_code) { joins(:lots).merge(Lot.by_okpd_code(okpd_code)) }
end

class Lot < ActiveRecord::Base
  belongs_to :customer, class_name: :Organization, inverse_of: :lots
  has_many :lot_items
  scope :by_okpd_code, ->(okpd_code) { joins(:lot_items).merge(LotItem.by_okpd_code(okpd_code)) }
end

class LotItem < ActiveRecord::Base
  belongs_to :lot
  belongs_to :okpd
  scope :by_okpd_code, ->(okpd_code) { joins(:okpd).where(tree_classifiers: { code: okpd_code } ) }
end

class TreeClassifier < ActiveRecord::Base
end

class Okpd < TreeClassifier
end
```
### Expected behavior

I use the scope `by_okpd_code` of `Organization`

So I should get the following sequenced SQL:

>  SELECT "organizations".\* FROM "organizations" INNER JOIN "lots" ON "lots"."customer_id" = "organizations"."id" LEFT OUTER JOIN "lot_items" ON "lot_items"."lot_id" = "lots"."id" LEFT OUTER JOIN "tree_classifiers" ON "tree_classifiers"."id" = "lot_items"."okpd_id" AND "tree_classifiers"."type" IN ('Okpd') WHERE "organizations"."id" IN (SELECT DISTINCT "lots"."customer_id" FROM "lots") AND "tree_classifiers"."code" = '60.77.43'

When
### Actual behavior

But I get the SQL, which I see left join on `tree_classifiers` **before** left join on `lot_items` as expected in:

> SELECT "organizations".\* FROM "organizations" INNER JOIN "lots" ON "lots"."customer_id" = "organizations"."id" LEFT OUTER JOIN "tree_classifiers" ON "tree_classifiers"."id" = "lot_items"."okpd_id" AND "tree_classifiers"."type" IN ('Okpd') LEFT OUTER JOIN "lot_items" ON "lot_items"."lot_id" = "lots"."id" WHERE "organizations"."id" IN (SELECT DISTINCT "lots"."customer_id" FROM "lots") AND "tree_classifiers"."code" = '60.77.43'

So when I issue the SQL request I get db exception (including the same in `psql` console)

But when I see the embedded scope I see the correct sequence:

> SELECT "lots".\* FROM "lots" INNER JOIN "lot_items" ON "lot_items"."lot_id" = "lots"."id" LEFT OUTER JOIN "tree_classifiers" ON "tree_classifiers"."id" = "lot_items"."okpd_id" AND "tree_classifiers"."type" IN ('Okpd') WHERE "tree_classifiers"."code" = '60.77.43'
### System configuration

**Rails version**: `4.2.6`

**Ruby version**: `ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]`

**Comment by MaxLap** [comment]
This is actually the same issue as #16140. 

As I cannot find a rationale for this behavior, I'm actually working on a pull request to solve this issue, but I found other issues along the way so it's taking a bit of extra time.

**Comment by Anagha22** [comment]
Hello 
I am a newbie.I would like to give a try in fixing this issue.Can I start working on it?

**Comment by prakashmurthy** [comment]
Created an [executable test case](https://github.com/prakashmurthy/executable_test_cases/blob/9653903c83730fbc06123bfac964f07308c9450c/embedded_model_error.rb) to reproduce this behavior with `postgresql`. 

The script fails on Rails 5.x and rails 4.x with the following error trace:

<details>
   <summary>error stack trace</summary>
```ruby
  1) Error:
BugTest#test_chained_merge_with_joins:
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "lot_items"
LINE 1: ...N "tree_classifiers" ON "tree_classifiers"."id" = "lot_items...
                                                             ^
: SELECT "organizations".* FROM "organizations" INNER JOIN "lots" ON "lots"."customer_id" = "organizations"."id" LEFT OUTER JOIN "tree_classifiers" ON "tree_classifiers"."id" = "lot_items"."okpd_id" AND "tree_classifiers"."type" IN ('Okpd') LEFT OUTER JOIN "lot_items" ON "lot_items"."lot_id" = "lots"."id" WHERE "tree_classifiers"."code" = $1
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:604:in `async_exec'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:604:in `block in exec_no_cache'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:586:in `block in log'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:579:in `log'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:604:in `exec_no_cache'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:593:in `execute_and_clear'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:103:in `exec_query'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:374:in `select'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:42:in `select_all'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:70:in `select_all'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/querying.rb:39:in `find_by_sql'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/relation.rb:689:in `exec_queries'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/relation.rb:573:in `load'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/relation.rb:258:in `records'
    /Users/prakash/.rvm/gems/ruby-2.3.1/bundler/gems/rails-0d2f79b0ca61/activerecord/lib/active_record/relation.rb:655:in `=='

1 runs, 1 assertions, 0 failures, 1 errors, 0 skips
```
</details>

The script works fine with `sqlite3` database adapter.

**Comment by maclover7** [comment]
Hi! Sorry for the delay in getting to this ticket.

- @Anagha22 if you would like to take a stab at solving this ticket, go right ahead! Always good to get new contributors involved.
- @prakashmurthy: can reproduce the error you are seeing with Postgres. My reproduction script [here]() appears to be working fine with SQLite.

Going to investigate this further. If anyone else would like to take a look here, please do! :)

**Comment by maclover7** [comment]
Hmm, running into something a little strange. Through when `exec_query` is called (and the SQL query is handed off to the actual SQLite3 / PG libraries), both the PG and SQLite3 adapters have the same exact SQL query and binds being passed through... 😬

**Comment by majioa** [comment]
@maclover7  had you trt to execute the pure SQL generated by merges in `psql`?

**Comment by maclover7** [comment]
Basically, I ran the reproduction scripts, and I had a `binding.pry` in `exec_query`. Every time a query was being run, I compared it between the two databases adapters. 😬

**Comment by majioa** [comment]
have you try also pg 0.17.1 driver for jruby?

**Comment by kamipo** [comment]
Fixed in #29413.
